src/Agent.Worker/Release/ContainerFetchEngine/FetchEngine.cs

```diff
@@ -0,0 +1,358 @@
+    /// </summary>
+    public abstract class FetchEngine : IDisposable
+    {
+
+        public IReleaseFileSystemManager FileSystemManager { private get; set; }
+        protected IContainerProvider Provider { get; }
+
+        protected FetchEngine(
+            IContainerProvider containerProvider,
+            string rootItemPath,
+            string rootDestinationDir)
+        {
+            RootItemPath = rootItemPath;
+            RootDestinationDir = rootDestinationDir;
+            Provider = containerProvider;
+
+            ContainerFetchEngineOptions = new ContainerFetchEngineOptions();
+            FileSystemManager = new ReleaseFileSystemManager();
+            ExecutionLogger = new NullExecutionLogger();
+        }
+
+        public void Dispose()
+        {
+            if (_downloadedFiles > 0)
+            {
+                ExecutionLogger.Output(StringUtil.Loc("RMDownloadComplete"));
+            }
+
+            LogStatistics();
+        }
+
+        protected async Task FetchItemsAsync(IEnumerable<ContainerItem> containerItems)
+        {
+            var itemsToDownload = new List<ContainerItem>();
+
+            foreach (ContainerItem item in containerItems)
+            {
+                if (item.ItemType == ItemType.Folder)
+                {
+                    string localDirectory = ConvertToLocalPath(item);
+                    FileSystemManager.EnsureDirectoryExists(localDirectory);
+                }
+                else if (item.ItemType == ItemType.File)
+                {
+                    string localPath = ConvertToLocalPath(item);
+
+                    ExecutionLogger.Info(StringUtil.Loc("RMCopyingFile", item.Path, localPath));
+
+                    _totalFiles++;
+
+                    if (item.FileLength == 0)
+                    {
+                        CreateEmptyFile(localPath);
+                        _newEmptyFiles++;
+                    }
+                    else
+                    {
+                        itemsToDownload.Add(item);
+                    }
+                }
+                else
+                {
+                    throw new NotSupportedException(StringUtil.Loc("RMContainerItemNotSupported", item.ItemType));
+                }
+            }
+
+            if (itemsToDownload.Count > 0)
+            {
+                // Used to limit the number of concurrent downloads.
+                SemaphoreSlim downloadThrottle = new SemaphoreSlim(ContainerFetchEngineOptions.ParallelDownloadLimit);
+                Stopwatch watch = Stopwatch.StartNew();
+                LinkedList<Task> remainingDownloads = new LinkedList<Task>();
+
+                foreach (ContainerItem ticketedItem in itemsToDownload)
+                {
+                    _bytesDownloaded += ticketedItem.FileLength;
+                    Task downloadTask = DownloadItemAsync(downloadThrottle, ticketedItem);
```

Comment:
will you run out of thread in threadpool? since you create a downloadTask for each item.